### ISSUE TITLE
coreaudio doesn't handle files with high bitrate

### ISSUE THREAD

**Issue by lochlainn** [cmus/cmus#912]
When I try to play songs with particularly high bitrate (over 2000kbps), coreaudio simply generates noise as output. 
Here's the info on the file I was testing:
```
Input #0, flac, from 'Daughters - You Won't Get What You Want - 01 City Song.flac':
  Metadata:
    TITLE           : City Song
    ARTIST          : Daughters
    DATE            : 2018
    COMMENT         : Visit http://daughters.bandcamp.com
    ALBUM           : You Won't Get What You Want
    track           : 1
    album_artist    : Daughters
  Duration: 00:05:55.74, start: 0.000000, bitrate: 2755 kb/s
    Stream #0:0: Audio: flac, 88200 Hz, stereo, s32 (24 bit)
    Stream #0:1: Video: mjpeg, yuvj444p(pc, bt470bg/unknown/unknown), 700x700 [SAR 72:72 DAR 1:1], 90k tbr, 90k tbn, 90k tbc
    Metadata:
      comment         : Cover (front)
```
---

1) `uname -a`:
```
$ uname -a
Darwin loch-pro 18.6.0 Darwin Kernel Version 18.6.0: Thu Apr 25 23:16:27 PDT 2019; root:xnu-4903.261.4~2/RELEASE_X86_64 x86_64
```

2) `cmus --version`:
```
$ cmus --version
cmus v2.8.0
Copyright 2004-2006 Timo Hirvonen
Copyright 2008-2016 Various Authors
```

3) `cmus --plugins`:
```
Input Plugins: /usr/local/Cellar/cmus/2.8.0/lib/cmus/ip
  mad:
    Priority: 55
    File Types: mp3 mp2
    MIME Types: audio/mpeg audio/x-mp3 audio/x-mpeg
  wav:
    Priority: 50
    File Types: wav
    MIME Types:
  mp4:
    Priority: 50
    File Types: mp4 m4a m4b
    MIME Types:
  cue:
    Priority: 50
    File Types:
    MIME Types: application/x-cue
  aac:
    Priority: 50
    File Types: aac
    MIME Types: audio/aac audio/aacp
  flac:
    Priority: 50
    File Types: flac fla
    MIME Types:
  vorbis:
    Priority: 50
    File Types: ogg oga ogx
    MIME Types: application/ogg audio/x-ogg

Output Plugins: /usr/local/Cellar/cmus/2.8.0/lib/cmus/op
  coreaudio
  ao

```

**Comment by tomty89** [comment]
Please test with a WAV file decoded from the FLAC file (with `flac -d`, which preserves the sample spec) and report back if the issue also occurs.

**Comment by tomty89** [comment]
Also, if you can, test with cmus git master. There are fixes for the flac input plugin, which might be relevant.

**Comment by lochlainn** [comment]
I tried again on git master with a WAV, but I'm still having the same issue.

**Comment by lochlainn** [comment]
I tried with a decoded WAV version of the same file, on cmus git master, and ran into the same issue. Do I need a newer flac? I'm using 1.3.2

**Comment by tomty89** [comment]
I think it has something to do with the fact that the coreaudio output plugin would bump the frame size to a power of 2 integer (probably 8) when the actual frame size is 6 with 24-bit stereo file (as the flac input plugin output "packed" signal; also so is coreaudio informed about the signal).

See if an "unpacked" wav version of the file triggers the problem:
```
ffmpeg -i input.flac -c:a pcm_s32le -ac 2 -ar 88200 output.wav
```

If it doesn't, I can try come up with a patch for you to test.

**Comment by lochlainn** [comment]
You're right, creating an unpacked wav causes the file to play properly. Is that behavior inherent to coreaudio?

**Comment by tomty89** [comment]
Never mind. I didn't realize that the ring buffer read/write code is all entangled with this silly power of 2 idea (so it seems; I don't exactly know what the big/small mask is doing).

To fix this we probably need a more general / less silly approach of ring buffer (array of pointers or double pointers, idk, but instead of this awful masking / first half second half approach).

**Comment by lochlainn** [comment]
Strangely enough, I no longer have this issue. I don't know if flac changed or coreaudio changed, but everything plays properly. I'll close this issue since nobody else is having it.

**Comment by lochlainn** [comment]
I've started having this issue again. Adjusting the volume on FLAC files with a high bitrate causes bursts of static.

If I wanted to try to fix this, where would I start? @tomty89 would you mind pointing me in the right direction on the ring buffer in the flac plugin?

**Comment by tomty89** [comment]
@lochlainn start from isolating/identifying the problem. `Adjusting the volume on FLAC files with a high bitrate causes bursts of static.` at least sounds entirely different from / unrelated to `coreaudio simply generates noise as output`. Also `high bitrate` is vague. Test with files of different specs and see exactly what kind of spec triggers (e.g. start from testing with files of the same format and sampling rate but only different bit depth; a.k.a. "fair test(s)").

**Comment by tomty89** [comment]
You are welcome to try this by the way: https://github.com/cmus/cmus/pull/992

**Comment by lochlainn** [comment]
Just pasting an update to make my comments more visible, after trying 995f754 I was able to play and pause a 24-bit flac track normally, but adjusting the softvol caused significant static and distortion. You can find a debug log [here](https://github.com/cmus/cmus/pull/992#issuecomment-801198136).

I have a lot more free time now, so I'm thinking I'll take a whack at this. @tomty89 is there anywhere you suggest I start looking to figure out the softvol issue?

**Comment by tidux** [comment]
For anyone who found this ticket via a search engine, `:set softvol=false` and managing volume at the OS level is the temporary workaround.

**Comment by Allavaz** [comment]
On my machine I also need to `:set output_plugin=ao`, `coreaudio` produces static noise even with `softvol` set to  `false`.

**Comment by dertuxmalwieder** [comment]
I can confirm that `:set output_plugin=ao` fixes the problem with WavPack files (I never had an issue with FLAC files).

**Comment by cubicvoid** [comment]
I was also getting terrible noise artifacts on (24-bit, 96k sample rate) flac on mac, constant with or without volume changes, also resolved by `:set output_plugin=ao`.

Is this potentially fixable in the coreaudio plugin? (Sounds like it would need some changes to the ring buffer implementation?) I have some macs to test on and would be willing to put some time into it.
